### DISCUSSION TITLE
Idiomatically passing pins used for both input and output

### DISCUSSION THREAD

**Discussion by djmdjm** [mciantyre/teensy4-rs#136] [Q&A] [answered]
I'm pretty new to Rust, so apologies in advance if this should be obvious.

What's the right way to pass a generic pin to a function (or via a struct) such that the pin can be switched between input and output later?

Background: I want to use the [HD44780 lcd module](https://docs.rs/lcd/latest/lcd/) and this requires filling out a context struct and implementing [a trait](https://docs.rs/lcd/latest/lcd/trait.Hardware.html) that uses the hardware resources referenced in the context struct to read and write from the device.

I can't figure out how to represent generic pins in the context struct in a way that will allow me to switch them between inputs and outputs. `iomuxc::ErasedPad` seems to be along the lines of what I need, but I can't figure out how to wrangle it back into something that I can use as an input or output without specifying the exact underlying pin or pad.

At the moment I have something like:

```rust
// Hardware struct for HD44780 character LCD
pub struct LcdHw<'a> {
    d4: &'a iomuxc::ErasedPad,
    d5: &'a iomuxc::ErasedPad,
    d6: &'a iomuxc::ErasedPad,
    d7: &'a iomuxc::ErasedPad,
    gpio: &'a gpio::Port<2>,
    // ...
}

impl<'a> lcd::Hardware for LcdHw<'a> {
    fn rw(&mut self, bit: bool) {
        if bit {
            // Configure d4..d8 as output
        } else {
            // Configure d4..d8 as input
        }
    }
    // ...
} 
```
Any clues?

**Comment by mciantyre** [comment, marked as answer]
One issue is that the erased pads don't know how to become a GPIO input / output. So even though we have those erased pad references, we're still not equipped to tell the GPIO port our intended direction or output state. The GPIO port normally depends on the strong type information to understand how to configure an input / output. That adds type complexity.

The solutions are definitely non-obvious. I'll provide [two options](https://github.com/mciantyre/teensy4-rs/commit/9fccaa5f34168c25297abfb786df2f34ff607e9d) that are possible in today's BSP. The commit message describes the options at a higher level. Haven't tested these beyond a type check, but I hope these can get you started.

I'm happy to answer follow-ons here, and discuss other designs that may be simpler for Rust newcomers.

**Comment by djmdjm** [comment]
Thanks for the quick and comprehensive response (and sorry for my being slow to get back to it). This gives me heaps to go on!
